src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs

```diff
@@ -494,6 +503,42 @@ protected Task<ImmutableArray<Document>> FindDocumentsWithDeconstructionAsync(Pr
             return locations.ToImmutableAndFree();
         }
 
+        protected async Task<ImmutableArray<ReferenceLocation>> FindReferencesInAwaitExpressionAsync(
+            ISymbol symbol,
+            Document document,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            var syntaxTreeInfo = await SyntaxTreeIndex.GetIndexAsync(document, cancellationToken).ConfigureAwait(false);
+            if (!syntaxTreeInfo.ContainsAwait)
+            {
+                return ImmutableArray<ReferenceLocation>.Empty;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var semanticFacts = document.GetLanguageService<ISemanticFactsService>();
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var locations = ArrayBuilder<ReferenceLocation>.GetInstance();
+
+            var originalUnreducedSymbolDefinition = symbol.GetOriginalUnreducedDefinition();
+
+            foreach (var node in syntaxRoot.DescendantNodesAndSelf())
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var awaitExpressionMethod = semanticFacts.GetAwaitExpressionMethod(semanticModel, node);
+
+                if (Matches(awaitExpressionMethod, originalUnreducedSymbolDefinition))
+                {
+                    var location = node.GetFirstToken().GetLocation();
+                    locations.Add(new ReferenceLocation(
+                        document, alias: null, location, isImplicit: true, isWrittenTo: false, CandidateReason.None));
+                }
+            }
+
+            return locations.ToImmutableAndFree();
+        }
```

Comment:
these methods feel like there's a lot of duplication (in the overall logic, even though individual bits seem different).  I'm on the fence if we should extract out a common helper...

Reply:
Thanks. I think the factorings you suggested are worth it, as I'll probably add support for `Dispose` in `using` and `Add` in collection initializers next.